Add TabPFGen Data Synthesizer Extension

sebhaan opened the pull request:
## Summary
This PR adds a new extension integrating [TabPFGen](https://github.com/sebhaan/TabPFGen) for synthetic tabular data generation.

## Impact
This extension enables TabPFN users to:

- Automatically balance imbalanced datasets using TabPFGen's latest features
- Generate high-quality synthetic tabular data with zero additional training
- Assess synthetic data quality with a suite of metrics and visualisations
- Integrate seamlessly with existing TabPFN workflows

Key Features:

- Energy-based synthetic data generation
- Support for both classification and regression tasks
- Automatic dataset balancing for imbalanced classes
- Class-balanced sampling option
- Comprehensive visualization tools
- Built on TabPFN transformer architecture
- No additional training required

## Installation 

```bash
# Install specific extension
pip install "tabpfn-extensions[tabpfgen_datasynthesizer]"

# Or install all extensions  
pip install "tabpfn-extensions[all]"
```

## Usage Examples

The extension provides multiple usage patterns, see :
- /tabpfn-extensions/src/tabpfn_extensions/tabpfgen_datasynthesizer/README.md 
- examples/tabpfn_datasynthesizer/

Basic classification dataset example:
```python
from tabpfn_extensions.tabpfgen_datasynthesizer import TabPFNDataSynthesizer
from sklearn.datasets import load_breast_cancer

# Load example data
X, y = load_breast_cancer(return_X_y=True)

# Initialize synthesizer
synthesizer = TabPFNDataSynthesizer(n_sgld_steps=500)

# Generate synthetic classification data
X_synth, y_synth = synthesizer.generate_classification(
    X, y, 
    n_samples=100,
    balance_classes=True,  # Only balances synthetic samples
    visualize=True         # TabPFGen's built-in visualization
)
```

## Project Structure

```
tabpfn-extensions/
├── src/tabpfn_extensions/tabpfgen_datasynthesizer/
│   ├── __init__.py             
│   ├── tabpfgen_wrapper.py   # TabPFNDataSynthesizer: Wrapper around TabPFGen functionality
│   └── utils.py                          #  Data validation, preparation, and quality assessment   
│   ├── README.md             
├── examples/tabpfn_datasynthesizer/
│   ├── basic_classification_example.py
│   ├── dataset_balancing_demo.py       
│   ├── basic_regression_example.py  
│   ├── README.md         
└── tests/
    └── test_tabpfgen_datasynthesizer.py 
```

## Checklist

- [x]  Follows project structure conventions
- [x]  All tests passing (pytest tests/test_tabpfgen_datasynthesizer.py -v)
- [x]  Examples working and documented
- [x]  Dependencies configured in pyproject.toml
- [x]  No breaking changes to existing code
- [x]  Detailed documentation

## Dependencies

Added to pyproject.toml: tabpfgen>=0.1.3 in both extension and [all] groups

## Documentation

- Detailed README with installation and usage examples
- Three detailed examples covering different use cases
- API documentation for all classes and methods
- Troubleshooting guide and performance tips

sebhaan committed:
Add main extension script for tabpfgen

sebhaan committed:
Add utils and readme to tabpfgen extension package

sebhaan committed:
Add examples for tabpfgen extension

sebhaan committed:
add test functions for tabpfgen extension

sebhaan committed:
Update pyproject file to include tabpfgen extension

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/PriorLabs/tabpfn-extensions?pullRequest=83) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/PriorLabs/tabpfn-extensions?pullRequest=83) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/PriorLabs/tabpfn-extensions?pullRequest=83) it.</sub>

noahho commented:
Hi @sebhaan, thanks again for this excellent contribution integrating `TabPFGen`!

To get this merged, please address the following:

**1. Critical CI Failures (Must Address First):**

* **Ruff Linting:** Please fix the linting errors by running `ruff format .` and `ruff check . --fix` locally and committing the changes.
* **Python 3.9 Dependency Conflict:** `tabpfgen 0.1.3` requires Python >=3.11, which conflicts with `tabpfn-extensions`'s Python 3.9 testing for the `[all]` extra. The error is:
    ```
    No solution found when resolving dependencies:
      ╰─▶ Because the current Python version (3.9.22) does not satisfy
          Python>=3.10 and tabpfgen==0.1.3 depends on Python>=3.11, we can
          conclude that tabpfgen==0.1.3 cannot be used.
          ...
          And because only tabpfn-extensions[all]==0.1.0 is available and you
          require tabpfn-extensions[all], we can conclude that your requirements
          are unsatisfiable.
    ```
    Please decide on a strategy:
    * **Option A (if feasible): Update `tabpfgen` for Python 3.9 Compatibility.**
        * Since you built and maintain `tabpfgen`, could you investigate if `tabpfgen` can be modified to support Python 3.9? This might involve checking for Python 3.10/3.11-specific syntax or dependencies.
        * If feasible, releasing a new `tabpfgen` version (e.g., `0.1.4`) that supports Python 3.9+ would allow this PR to update its dependency and keep `tabpfgen` in the `[all]` extra.
    * **Option B (If Option A is not practical): Isolate This Extension's Python Requirement.**
        * In `tabpfn-extensions`' `pyproject.toml`, remove `tabpfgen` from the `project.optional-dependencies.all` list (it should remain in `project.optional-dependencies.tabpfgen_datasynthesizer`).
        * Ensure tests for this extension are correctly skipped on Python <3.11 (the existing `TABPFGEN_AVAILABLE` flag should facilitate this if `tabpfgen` isn't installed in those environments).
        * Document in this extension's README that it requires Python >=3.11.
* **License/CLA:** Ensure the pending CLA status is resolved by signing CLA (this is standard for OS projects and we just want to double check; contribution is Apache 2.0 so this is a formality; also could only affect the code in this repo of course)

**2. Code:**

* **READMEs:**
    * Consolidate primary documentation into `src/tabpfn_extensions/tabpfgen_datasynthesizer/README.md` (as suggested by @noahho).
    * Significantly trim `examples/tabpfgen_datasynthesizer/README.md` to a brief purpose statement and link to the main extension README.
    * Relocate general installation instructions for the extension (e.g., `pip install ...[tabpfgen_datasynthesizer]`) to the top-level `PriorLabs/tabpfn-extensions` README. The extension's own README should then detail its specific Python version requirements (per point 1B above, if chosen).

sebhaan committed:
Fix linting issues with ruff format and ruff check --fix

sebhaan committed:
Consolidated example README into src README and rm of example README

sebhaan committed:
Add tabpfgen_datasynthesizer extension and installation instruction

sebhaan committed:
Remove tabpfgen dependency from all

sebhaan committed:
Add python version compatibility check

sebhaan committed:
Update test using tabpfgen_datasynthesizer extension with Python 3.10+ requirement isolation

sebhaan commented:
Thank you for the detailed feedback. I've addressed the critical issues as follows:

## 1. Critical CI Failures - RESOLVED 

Ruff Linting: - FIXED
- Ran `ruff format .` and `ruff check . --fix` locally
- All linting issues have been resolved and changes committed

Python 3.9 Dependency Conflict: - FIXED via Option B
- Updated `tabpfgen` to version 0.1.4 with relaxed Python requirement from `>=3.11` to `>=3.10`
- Implemented Option B as suggested:
  - Removed `tabpfgen` from the `[all]` extra in `pyproject.toml`
  - `tabpfgen` remains available via `[tabpfgen_datasynthesizer]` extra only
  - Added Python version compatibility check in the extension code
  - Tests are properly skipped on Python <3.10 and EXTENSION_AVAILABLE flag

This approach ensures:
- Main `tabpfn-extensions` package maintains Python 3.9+ compatibility
- Users who need this specific extension can install it with the appropriate Python version
- CI passes on all supported Python versions

License/CLA:
- Signed the CLA as requested

## 2. Code Structure - RESOLVED 

README Consolidation: - COMPLETED
- Consolidated the example README into `src/tabpfn_extensions/tabpfgen_datasynthesizer/README.md`
- Removed the redundant `examples/tabpfgen_datasynthesizer/README.md`
- Updated top-level README with:
  - `tabpfgen_datasynthesizer` extension listing
  - Installation instructions: `pip install tabpfn-extensions[tabpfgen_datasynthesizer]`
  - Clear Python >=3.10 requirement documentation

### Summary of Changes Made:

1. ✅ Fixed all ruff linting issues
2. ✅ Updated `tabpfgen` dependency to version 0.1.4 (Python >=3.10)
3. ✅ Implemented Option B dependency isolation strategy
4. ✅ Consolidated and restructured all README files
5. ✅ Added proper Python version compatibility checks
6. ✅ CLA signed

sebhaan commented:
Thank you for the detailed feedback. I've addressed the critical issues as follows:

## 1. Critical CI Failures - RESOLVED ✅

Ruff Linting: ✅ FIXED
- Ran `ruff format .` and `ruff check . --fix` locally
- All linting issues have been resolved and changes committed

Python 3.9 Dependency Conflict: ✅ FIXED via Option B
- Updated `tabpfgen` to version 0.1.4 with relaxed Python requirement from `>=3.11` to `>=3.10`
- Implemented Option B as suggested:
  - Removed `tabpfgen` from the `[all]` extra in `pyproject.toml`
  - `tabpfgen` remains available via `[tabpfgen_datasynthesizer]` extra only
  - Added Python version compatibility check in the extension code
  - Tests are properly skipped on Python <3.10 and EXTENSION_AVAILABLE flag

This approach ensures:
- Main `tabpfn-extensions` package maintains Python 3.9+ compatibility
- Users who need this specific extension can install it with the appropriate Python version
- CI passes on all supported Python versions

License/CLA:
- Signed the CLA as requested

## 2. Code Structure - RESOLVED ✅

README Consolidation: ✅ COMPLETED
- Consolidated the example README into `src/tabpfn_extensions/tabpfgen_datasynthesizer/README.md`
- Removed the redundant `examples/tabpfgen_datasynthesizer/README.md`
- Updated top-level README with:
  - `tabpfgen_datasynthesizer` extension listing
  - Installation instructions: `pip install tabpfn-extensions[tabpfgen_datasynthesizer]`
  - Clear Python >=3.10 requirement documentation

### Summary of Changes Made:

1. ✅ Fixed all ruff linting issues
2. ✅ Updated `tabpfgen` dependency to version 0.1.4 (Python >=3.10)
3. ✅ Implemented Option B dependency isolation strategy
4. ✅ Consolidated and restructured all README files
5. ✅ Added proper Python version compatibility checks
6. ✅ CLA signed
